homework/index.js

```diff
@@ -30,17 +34,110 @@
     return elem;
   }
 
+  function renderError(error) {
+    const body = document.body;
+    body.innerHTML = '';
+    createAndAppend('h1', body, { text: error.message, class: `alert-error` });
+  }
+
+  function addRow(tbody, label, value) {
+    const tr = createAndAppend('tr', tbody);
+    createAndAppend('td', tr, { text: `${label}:`, class: 'label' });
+    createAndAppend('td', tr, { text: value });
+  }
+
+  function renderRepos(repoName, ul) {
+    fetchJSON(`${HYF_REPOS_URL}`, (err, repositories) => {
+      if (err) {
+        renderError(err);
+        return;
+      }
+      ul.innerHTML = '';
+      repositories.forEach(repo => {
+        if (repo.name === repoName) {
+          const li = createAndAppend('li', ul);
+          const table = createAndAppend('table', li);
+          const tbody = createAndAppend('tbody', table);
+          const tr1 = createAndAppend('tr', tbody);
+          createAndAppend('td', tr1, { text: `Name:`, class: 'label' });
+          const a = createAndAppend('a', tr1, { href: `${repo.html_url}` });
+          createAndAppend('td', a, { text: `${repo.name}` });
+          addRow(tbody, 'Description', `${repo.description}`);
+          addRow(tbody, 'Forks', `${repo.forks}`);
+          addRow(tbody, 'Updated', `${repo.updated_at.slice(0, 10)}`);
+        }
+      });
+    });
+  }
+
+  function renderContributions(repoName, ul) {
+    fetchJSON(
+      `https://api.github.com/repos/HackYourFuture/${repoName}/contributors`,
+      (err, repositories) => {
+        if (err) {
+          renderError(err);
+          return;
+        }
+        ul.innerHTML = '';
+        createAndAppend('h3', ul, { text: `Contributors:` });
+        repositories.forEach(repo => {
+          const a = createAndAppend('a', ul, { href: `${repo.html_url}` });
+          const li = createAndAppend('li', a);
+          const table = createAndAppend('table', li);
+          const tbody = createAndAppend('tbody', table);
+          const tr1 = createAndAppend('tr', tbody);
+          createAndAppend('img', tr1, { src: `${repo.avatar_url}` });
+          createAndAppend('td', tr1, { text: `${repo.login}` });
+          createAndAppend('td', tr1, { text: `${repo.contributions}` });
+        });
+      },
+    );
+  }
+
+  function createOptionElements(repositories, select) {
+    createAndAppend('option', select, { text: 'Select a repository', disabled: `disabled` });
+    repositories
+      .sort((a, b) => a.name.localeCompare(b.name))
+      .forEach(repo => {
+        createAndAppend('option', select, { text: repo.name, value: repo.name });
```

Comment:
I spent a long time trying to make this work. I finally figured it out for renderRepos function but I still could not remove the forEach loop in renderContributions function. Was that necessary? Is there a way of doing this without looping through contributors?

Reply:
You need the `.forEach()` for in `renderContributors` because you need to create a list item for each of contributor in the `contributors` array.

But perhaps you confused yourself by using the parameter name `repositories` for the `fetchJSON()` function inside `renderContributors()`. Your are not fetching `repositories`, you are fetching `contributors`.

```js
function renderContributions(repo, ul) {
  fetchJSON(repo.contributors_url, (err, contributors) => {
    if (err) {
      renderError(err);
      return;
    }
    ul.innerHTML = '';
    createAndAppend('h3', ul, { text: `Contributors:` });
    contributors.forEach(contributor => {
      const a = createAndAppend('a', ul, { href: contributor.html_url });
      ...
    });
  });
}
```

Reply:
Yes I understand now thanks